[FLINK-8689][table]Add runtime support of distinct filter using MapView for codegen

walterddr opened the pull request:
## What is the purpose of the change

- Adding in runtime support using distinct aggregation delegate to support distinct filtering using MapView. This is only fixing the currently broken over-window aggregate with distinct filter, however this change is meant to be used by other distinct operations on datastream such as group window aggregate, group aggregate, etc.


## Brief change log

  - Adding a new DistinctAggDelegateFunction and DistinctAccumulator that encapsulates any real aggregate function.
  - change codegen to specifically generate distinct filter before invoking the actual aggregate function.
  - adding in more codegen specifically for using delegates on merge and reset.

## Verifying this change

This change added tests and can be verified as follows:

  - Added over-window unit-test to verify generated plan before codegen.
  - Added integration tests for testing distinct over-window aggregate end-to-end.

## Does this pull request potentially affect one of the following parts:

no

## Documentation

this does not expose and additional external facing functionality, which will come in separated PR.

fhueske commented:
Hi @walterddr, thanks for working on this and opening the PR! 
I'm quite busy right now, but will try to have a look at the changes next week.

Thank you, Fabian

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/AggregationCodeGenerator.scala:
```diff
@@ -105,11 +106,32 @@ class AggregationCodeGenerator(
+        agg
+    }
+
+    val realAggTypes = aggregates.map {
```
Can be replaced by `realAggregates.map(_.getClass.getName)`

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/DistinctAggDelegateFunction.scala:
```diff
@@ -0,0 +1,122 @@
+import org.apache.flink.table.api.dataview.MapView
+import org.apache.flink.table.dataview.MapViewTypeInfo
+import org.apache.flink.table.functions.AggregateFunction
+
```
Add comments for this class and functions.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/DistinctAggDelegateFunction.scala:
```diff
@@ -0,0 +1,122 @@
+import org.apache.flink.table.dataview.MapViewTypeInfo
+import org.apache.flink.table.functions.AggregateFunction
+
+class DistinctAccumulator[E, ACC] (var mapView: MapView[E, Integer], var realAcc: ACC) {
```
Remove blanks between `[E, ACC]` and `(var mapView`

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/DistinctAggDelegateFunction.scala:
```diff
@@ -0,0 +1,122 @@
+    }
+  }
+
+  def accumulate(acc: DistinctAccumulator[E, ACC], element: E, count: Int): Boolean = {
```
Add a corresponding retract function.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/DistinctAggDelegateFunction.scala:
```diff
@@ -0,0 +1,122 @@
+      new MapViewTypeInfo[E, Integer](
+        elementTypeInfo.asInstanceOf[TypeInformation[E]], BasicTypeInfo.INT_TYPE_INFO)))
+    pojoFields.add(new PojoField(clazz.getDeclaredField("realAcc"),
+      realAgg.getAccumulatorType))
```
`getAccumulatorType ` may return null if has not been overrided. In this case, accumulator type should be inferred. see `getAccumulatorTypeOfAggregateFunction()` in `UserDefinedFunctionUtils`.

walterddr replied:
Good catch!

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/DistinctAggDelegateFunction.scala:
```diff
@@ -0,0 +1,122 @@
+        elementTypeInfo.asInstanceOf[TypeInformation[E]], BasicTypeInfo.INT_TYPE_INFO)))
+    pojoFields.add(new PojoField(clazz.getDeclaredField("realAcc"),
+      realAgg.getAccumulatorType))
+    new PojoTypeInfo[DistinctAccumulator[E, ACC]](clazz, pojoFields)
```
Currently, `MapView` and `ListView` only supported at first level of accumulators of Pojo type. In other words, if  
 realAcc also contains MapView, exceptions will be thrown. You can reproduce this exception by creating a test with `org.apache.flink.table.runtime.utils.JavaUserDefinedAggFunctions.CountDistinct(DISTINCT a)`.

walterddr replied:
Yes, I was actually aware of that as one of the limitation of the `DataView` API and will propose a change in another PR. At this point since this is solely used for distinct aggregation. I think explicitly catch and thrown and exception if `realAcc` contains Pojo type?

hequn8128 replied:
Make sense.

hequn8128 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/AggregateUtil.scala:
```diff
@@ -1393,6 +1393,21 @@ object AggregateUtil {
       }
+
+      // create distinct accumulator delegate
+      if (aggregateCall.isDistinct) {
```
It is better to move this logic into the upper match. Distinct is also a kind of aggregate whose fields should not be empty. Also, we can reuse some variables. What do you think?

walterddr replied:
Good point. Actually there's a very interesting question I've been thinking - This AggFunction literally is the distinct version of the  `CollectAggFunction`, however it cannot exist by itself unless chained with other `realAgg` functions --> this means we will need to chain `DistinctAggFunction` with `CollectAggFunction` which just dont make sense. I will try some other approach and see if they work.

hequn8128 replied:
Sql will be verified by calcite to exclude single DISTINCT during sql parse phase, so maybe we don't have to consider single distinct here.

hequn8128 commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/OverWindowITCase.scala:
```diff
@@ -50,6 +50,80 @@ class OverWindowITCase extends StreamingWithStateTestBase {
     (20L, 20, "Hello World"))
 
+  @Test
+  def testUnboundedDistinctGroupWindow(): Unit = {
```
Remove this test case. The distinct in this case is solved by `AggregateExpandDistinctAggregatesRule`

walterddr replied:
oops.. this was suppose to be added later in FLINK-8690 :-P

fhueske commented:
Hi @walterddr and @hequn8128, thanks for the PR, review, and discussions.

The current implementation with the `DistinctAggDelegateFunction` and accumulators takes the path of user-defined code although `DISTINCT` is a query feature that does not require user-code. 
Wouldn't it be easier if we would just add two parameters to `AggregationCodeGenerator.generateAggregations()` methods:

- distinctAggs: Array[Boolean]
- stateBackedDistinct: Option[Boolean]

and handle the distinct in the generated code? Given this information we can configure the required MapViews (also reusing them across multiple aggregation functions). Also we don't need to an aggregation function but access the MapView directly and check for distinct input or not.

This would mean a bit more implementation effort for the code-generation, but be the cleaner design because we do not need to wrap aggregation function and accumulators. It would avoid all problems with nested map views and make the planning code easier.

What do you think @walterddr, @hequn8128?

hequn8128 commented:
Cool! It can avoid the problems with nested map views and is more flexible.

walterddr commented:
I agree this design is much cleaner and easier to maintain later. I was hesitating to change the function signature of `generateAggregations()`.

I will try to introduce some common building blocks and handle implementation effort in codegen. I will imagine a lot of ITCases will be needed :-)

Thanks for the feedback @hequn8128 @fhueske

walterddr commented:
@hequn8128 @fhueske Thanks for the feedback. I have updated the diff to directly use DistinctAccumulator for filtering and modified the `generateAggregation` API. Please kindly take another look when you have time. I have resolved the issue of multiple layer dataview codegen. 

In terms of reusing same `DataView` for multiple distinct aggregations against the same field, I tried to incorporate but there are many assumptions with single mapping between `AggregateFunction`s and `Accumulator`s that's hard to deal with. I am planning to continue and improve on it in a separated JIRA, what do you think?

hequn8128 commented:
@walterddr Cool, thanks for your update. I will take a look during this weekend or early next week :-)

fhueske commented:
Yes, thanks for the update @walterddr! I'm travelling this week, but will try to have a look next week as well.

walterddr commented:
Thanks @hequn8128 for the prompt review. 
Are you suggesting we created the mapView parallel with the accumulator? The reason why I kept `DistinctAccumulator` is to act as a delegate to enclose the actual accumulator so that it can be passed around in the `accumulatorState` field without extending the arity.

I guess if we separate the mapView with the accumulator. I guess I can separately create another field in the `accumulatorState` `Row` to store the `mapView`(s)... This way it might be easier to handle the "reuse same mapView for multiple different distinct agg function" case as we discussed in the doc.

Another question is I was trying to reuse as much utility of dataview codegen as possible, as most of them are tightly coupled with the accumulators. I guess I can further refactor (which I already did quite a bit already).

Please let me know if that's what you had in mind @hequn8128 

--
Rong

hequn8128 commented:
Hi, @walterddr 
I was thinking we added some `MapView` as new member variables of the generated class, so we don't need to extend the arity of `accumulatorState `.  What do you think?

walterddr commented:
Hi @fhueske . Thanks for the review, all very good points. I will follow up with the next steps. Actually @hequn8128  and I had some discussions regarding the follow up in FLINK-8690 already and I created 2 different approaches. Please kindly take a look when you have time :-)

Best,
Rong

walterddr commented:
Thanks @fhueske, I had the exact same feeling. Just attaching `MapState` towards the back of the `Row` might be a current working solution for now, but will probably be nasty to maintain in the future. 

I am planning to go ahead and create an optimization ticket further down once I had FLINK-8690 completed. What do you think?

fhueske commented:
Thanks, that'd be great!

walterddr commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/OverWindowITCase.scala:
```diff
@@ -50,6 +50,155 @@ class OverWindowITCase extends StreamingWithStateTestBase {
+  }
+
+  @Test
+  def testRowTimeDistinctBoundedNonPartitionedRowsOver(): Unit = {
```
This test case was added to cover all over aggregate usage on distinct accumulator.

walterddr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/AggregationCodeGenerator.scala:
```diff
@@ -151,8 +157,15 @@ class AggregationCodeGenerator(
-    addReusableDataViews()
+    // get distinct filter of acc fields for each aggregate functions
+    val distinctAccType = s"${classOf[DistinctAccumulator[_, _]].getName}"
+    val isDistinctAggs = distinctAggs.map(_.nonEmpty)
```
Yes. You are right. Sorry I missed that. Just updated :-)

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/AggregationCodeGenerator.scala:
```diff
@@ -481,14 +622,34 @@ class AggregationCodeGenerator(
+               |      Object k = entry.getKey();
+               |      Long v = (Long) entry.getValue();
+               |      if (aDistinctAcc$i.add(k, v)) {
+               |        ${accTypes(i)} aAcc$i = (${accTypes(i)}) aDistinctAcc$i.getRealAcc();
```
Move this out of the while loop?

fhueske commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/DistinctAccumulator.scala:
```diff
@@ -0,0 +1,106 @@
+    }
+
+  def add(element: E): Boolean = {
+    if (element != null) {
```
Shouldn't it be up to the aggregation function how to handle `null` valued parameters?
I'd pass the null value on to the function instead of filtering it out. We'd need to wrap it in a `Row()` with a single field to distinguish the case of empty entry and null-valued entry.

walterddr replied:
Great catch @fhueske. Sorry I misunderstood what you meant previous regrading `null` value. This line was to disregard all `null` values. But as you mentioned, some operators might treat `null` value differently, e.g. `COUNT` aggregators. I added the handling and the tests as well.

fhueske commented on flink-libraries/flink-table/src/test/scala/org/apache/flink/table/runtime/stream/sql/OverWindowITCase.scala:
```diff
@@ -50,6 +50,155 @@ class OverWindowITCase extends StreamingWithStateTestBase {
+    val env = StreamExecutionEnvironment.getExecutionEnvironment
+    env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime)
+    env.setStateBackend(getStateBackend)
+    env.setParallelism(1)
```
should also work for the default parallelism (p=1 only important for proc time processing)

fhueske replied:
Actually, I was wrong on this one. Late elements are not deterministic handled if p > 1.
Will change it back.

fhueske replied:
Also, the event-time test is not required. We test the retract case also with BOUNDED OVER windows (rows the fall out of the window are retracted).

walterddr replied:
I think all other DISTINCT aggregates are tested against proctime so I was just trying to use this one to test out-of-order delivery, as well as test against rowtime. 

this also triggers `DistinctAccumulator.remove()` which was never used in other proctime test cases.

fhueske replied:
By default the LocalExecEnv chooses the parallelism is picked based on the number CPU cores. Are you running the tests in some kind of container?

walterddr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/AggregationCodeGenerator.scala:
```diff
@@ -151,8 +157,36 @@ class AggregationCodeGenerator(
+      case (isDistinctAgg, idx) => if (isDistinctAgg) {
+        val fieldIndex: Int = aggFields(idx)(0)
+        val mapViewTypeInfo = new MapViewTypeInfo(
+          physicalInputTypes(fieldIndex), BasicTypeInfo.LONG_TYPE_INFO)
```
at this moment it will disregard any null value. But as you mention this is not correct. Will address.

walterddr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/codegen/AggregationCodeGenerator.scala:
```diff
@@ -151,8 +157,36 @@ class AggregationCodeGenerator(
+          physicalInputTypes(fieldIndex), BasicTypeInfo.LONG_TYPE_INFO)
+        Seq(
+          MapViewSpec(
+            "distinctAgg" + idx + "_field" + fieldIndex,
```
+1

walterddr commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/aggfunctions/DistinctAccumulator.scala:
```diff
@@ -47,60 +51,51 @@ class DistinctAccumulator[E, ACC](var realAcc: ACC, var mapView: MapView[E, JLon
-        true
-      }
-    } else {
+    val wrappedElement = Row.of(element)
```
Thanks @fhueske for the insight. yeah I thought about that before the last commit but didn't go through with it, since we still need to construct the row of single element before passing to the distinct accumulator. But you are right, it will make future optimization easier
